### ISSUE TITLE
Get policies working on `main` based peer pods

### ISSUE THREAD

**Issue by stevenhorsman** [confidential-containers/cloud-api-adaptor#1874]
*As a* peer pods workload owner
*I want* to have be able to block certain endpoints
*so that* I can restrict the interactions that users can have with my workloads

*Description*
In https://github.com/confidential-containers/cloud-api-adaptor/pull/1607 Pradipta added policy based on the `CCv0` stream, however after some testing that @snir911 and I have done, we've found it doesn't work with the changed implementation on kata-containers `main` branch.

We should investigate this, try and get it working and ideally add the e2e test that were supposed to come following that PR.

*Acceptance Criteria*

*_Scenario: Block exec command*
*Given* a peer pods set-up including the "latest" kata-containers `main` codebase and a pod created, which includes the annotation `io.katacontainers.config.agent.policy:` which is set to the base64 versions of `allow-all-except-exec-process.rego`
*When* we run `kubectl exec <pod_name> -- ls`
*Then* The request is blocked with an error that includes: `failed to exec in container` and `Validation failed for data.policy.ExecProcess` 

*_Scenario: Fully permission policy*
*Given* a peer pods set-up including the "latest" kata-containers `main` codebase and a pod created, which includes the annotation `io.katacontainers.config.agent.policy:` which is set to the base64 versions of `allow-all.rego`
*When* we run `kubectl exec <pod_name> -- ls`
*Then* The request is successful

*_Scenario: Check default policy is permissive_*
*Given* a peer pods set-up including the "latest" kata-containers `main` codebase and a pod created, with no `io.katacontainers.config.agent.policy:` annotation
*When* we run `kubectl exec <pod_name> -- ls`
*Then* The request is successful

**Comment by stevenhorsman** [comment]
In our initial investigation it looks like there are potentially two problems:
1. It looks like the policy stuff should be set in https://github.com/kata-containers/kata-containers/blob/b218c4bc1096f56b62830f892ff516cea44a60b2/src/runtime/pkg/oci/utils.go#L952
and read from the ociSpec annotations, but there doesn't seem to be anything in the logs to show that is going through and we end up with no policy in the AgentConfig:
AgentConfig:{KernelModules:[] ContainerPipeSize:0 DialTimeout:0 LongLiveConn:true Debug:false Trace:false EnableDebugConsole:false Policy:}. We probably need some more logging here to debug
2. The SetPolicy command isn't invoked on the remote hypervisor: https://github.com/kata-containers/kata-containers/blob/cf372f41bf5a6e75b2f8aeb15038c2b68077e50c/src/runtime/virtcontainers/kata_agent.go#L767-L780 we will need a kata-containers change for this.

**Comment by stevenhorsman** [comment]
OK, I create a local branch which fixed part 2 and even though Policy wasn't logged in the `AgentConfig` that worked, so maybe it's missing for security reasons or something? I'll create the kata-containers PR and hopefully resolve this soon.

**Comment by stevenhorsman** [comment]
9881 has been merged now, so I've created https://github.com/confidential-containers/cloud-api-adaptor/pull/1876 to add the e2e tests for this

**Comment by mkulke** [comment]
How can we test that the policy works again? is it sufficient to deploy the kata-runtime using a recent version of the operator:

```bash
kubectl apply -k "github.com/confidential-containers/operator/config/release?ref=main"
kubectl apply -k "github.com/confidential-containers/operator/config/samples/ccruntime/peer-pods?ref=main"
```

**Comment by stevenhorsman** [comment]
> How can we test that the policy works again? is it sufficient to deploy the kata-runtime using a recent version of the operator:
> 
> ```shell
> kubectl apply -k "github.com/confidential-containers/operator/config/release?ref=main"
> kubectl apply -k "github.com/confidential-containers/operator/config/samples/ccruntime/peer-pods?ref=main"
> ```

Unfortunately I think not - the kata-container PR got merged on 20th June, but unfortunately we haven't had a successful kata-containers deployment payload since the day before. Unfortunately there is a problem with the arm runners :( 

The way I tested it was to log in to the worker and rebuild and install the kata shim manually, which isn't too bad, so I can give you a snippet if that helps?

**Comment by mkulke** [comment]
> The way I tested it was to log in to the worker and rebuild and install the kata shim manually, which isn't too bad, so I can give you a snippet if that helps?

yeah that would help, thx.

**Comment by snir911** [comment]
@mkulke IIRC i could deploy latest by using kata-deploy-ci amd64 specific image which i found in the registry (the multiarch indeed wasn't there):
for example:
`quay.io/kata-containers/kata-deploy-ci:kata-containers-e3e38738576682b35eafce7ecc5bb524884d4e21-amd64`
(BTW agent policy tested and worked perfectly :+1: )

**Comment by stevenhorsman** [comment]
Yeah, Snir's approach works if you don't mind tweaking the operator code. Otherwise the approach I did pre-merge to test it is to log into the worker node and then run:
```
# Install dependencies
sudo apt install make
sudo snap install go --classic
# Pull in the version of code you need, `main` should be fine now it has merged
git clone https://github.com/kata-containers/kata-containers.git
cd kata-containers/src/runtime
# Build and install shim
make $PWD/containerd-shim-kata-v2
sudo install containerd-shim-kata-v2 /usr/local/bin/containerd-shim-kata-remote-v2
```
So it's not too bad once you've worked out the command and pre-reqs

**Comment by stevenhorsman** [comment]
Working and tested in https://github.com/confidential-containers/cloud-api-adaptor/pull/1876
